Add sorting decorators (like old Redoc had)

lornajane opened the pull request:
Adds some of the sorting features that older versions of open source Redoc (v1 and v2) could apply at display time so that they can be used at any stage of the API lifecycle.

adamaltman commented on custom-plugin-decorators/sorting/README.md:
```diff
@@ -0,0 +1,94 @@
+Redocly CLI has some rules about sorting, that these decorators can help to satisfy:
+
+- [`tags-alphabetical`](https://redocly.com/docs/cli/rules/tags-alphabetical/) rule to error if the `tags` section isn't in alphabetical order by `name`.
+- [`required-props-first`](https://redocly.com/docs/cli/rules/tags-alphabetical/) rule to error if an object's properties don't put the required fields first.
```
This links to tags-alphabetical

adamaltman replied:
I don't think we have such a rule for `required-props-first` (or I can't find it).

lornajane replied:
Huh! I guess I should write it as well in that case! It was on my notes, but maybe it was a "todo"?

lornajane replied:
Added a `property-sort` rule

thesteve0 commented:
Does this also include a custom rule to specify the order expected by the linter?

lornajane commented:
I don't have a rule for the method ordering included, but I think we can add one.

adamaltman commented:
I agree that it should accept a config for method order (maybe it can have a default for zero config). I can see particular people wanting to go in some specific order though which might not be alphabetical (such as `DELETE` last).

thesteve0 commented:
Yeah for us we want GET, POST, PUT, DELETE

On Thu, Aug 22, 2024 at 6:56 PM Adam Altman ***@***.***>
wrote:

> I agree that it should accept a config for method order (maybe it can have
> a default for zero config). I can see particular people wanting to go in
> some specific order though which might not be alphabetical (such as DELETE
> last).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Redocly/redocly-cli-cookbook/pull/38#issuecomment-2306035708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKWZQ6264CWOD3P7NAFIALZS2JE3AVCNFSM6AAAAABMUK2YU6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBWGAZTKNZQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tatomyr commented on custom-plugin-decorators/sorting/sort-props-required.js:
```diff
@@ -0,0 +1,33 @@
+    Schema: {
+      leave(schema) {
+        if(schema.type == "object") {
+          const propList = Object.getOwnPropertyNames(schema.properties);
+          let newProps = {}
+
+          if(schema.required && schema.required.length > 0) {
+            const requiredList = schema.required;
+            // put the required items in first
+            requiredList.forEach((prop) => {
+              newProps[prop] = schema.properties[prop];
+            });
+
+            // now add anything that wasn't already added
+            propList.forEach((prop) => {
+              if(!newProps[prop]) {
+                newProps[prop] = schema.properties[prop];
+              }
+            });
+            schema.properties = newProps;
```
```suggestion
          const { properties = {}, required = [] } = schema;
          for (const prop in properties) {
            if (!required.includes(prop)) {
              const p = properties[prop];
              // Delete and add the property back to the object if it's not required, so it will be at the end of the object:
              delete properties[prop];
              properties[prop] = p;
            }
          }
```

lornajane replied:
This would work for this specific case, but I think as an overall pattern, the more verbose and readable approach of constructing a fresh array and replacing the old one is probably a more re-usable practice.

tatomyr commented on custom-plugin-decorators/sorting/sort-tags.js:
```diff
@@ -0,0 +1,16 @@
+  return {
+    TagList: {
+      leave(target) {
+        target.sort((a,b) => {
+          if (a.name < b.name) {
+            return -1;
+          }
+        });
```
```suggestion
        target.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
```

lornajane replied:
In favour of readability (this code is meant to be read and understood by all users) I don't think this is an improvement.

tatomyr reviewed (commented):
Left a few comments from the code style prospective.

adamaltman commented:
Shall we push this one through to completion? Seems powerful.

lornajane commented:
@adamaltman Definitely! Want to take another look? I added the extra rules we discussed.

@thesteve0 I added a `method-sort` rule that I hope will do what you need, let me know if you get chance to try it?

adamaltman commented on custom-plugin-decorators/sorting/README.md:
```diff
@@ -0,0 +1,130 @@
+When Redocly only made API documentation tools, the sorting changes were made as part of the docs build.
+But now we make more complex tools, and many API pipelines have more than just docs in them too - so these operations are more commonly done with a decorator to transform the API description; then it can be used with any tools.
+
+Redocly CLI has a [`tags-alphabetical`](https://redocly.com/docs/cli/rules/tags-alphabetical/) rule to error if the `tags` section isn't in alphabetical order by `name`.
```
```suggestion
Redocly CLI has a [`tags-alphabetical`](https://redocly.com/docs/cli/rules/tags-alphabetical) rule to error if the `tags` section isn't in alphabetical order by `name`.
```

adamaltman commented on custom-plugin-decorators/sorting/README.md:
```diff
@@ -0,0 +1,130 @@
+
+## References
+
+- [`tags-alphabetical' rule](https://redocly.com/docs/cli/rules/tags-alphabetical/)
```
```suggestion
- [`tags-alphabetical' rule](https://redocly.com/docs/cli/rules/tags-alphabetical)
```

adamaltman commented on custom-plugin-decorators/sorting/rule-sort-methods.js:
```diff
@@ -0,0 +1,36 @@
+    PathItem: {
+      enter(pathItem, ctx) {
+        // default method sort order, can be changed with an "order" param in config
+        let methodOrder = ["post", "patch", "put", "get", "delete"];
```
Should we add head, options?

lornajane replied:
I didn't on the basis that it's configurable so here's a very mainstream starter set and people can work on it from there. Although now you mention it, maybe "get" should be first!
